File: src/main/java/xyz/phanta/cbtweaker/util/CapabilityMerger.java
package xyz.phanta.cbtweaker.util;

import io.github.phantamanta44.libnine.util.helper.EnergyUtils;
import io.github.phantamanta44.libnine.util.helper.InventoryUtils;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.energy.CapabilityEnergy;
import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
import net.minecraftforge.fluids.capability.templates.FluidHandlerConcatenate;
import net.minecraftforge.items.CapabilityItemHandler;
import xyz.phanta.cbtweaker.CbtMod;

import javax.annotation.Nullable;
import java.util.*;

@FunctionalInterface
public interface CapabilityMerger<T> {

    Registry REGISTRY = new Registry();

    T merge(List<T> instances);

    class Registry {

        private final Map<Capability<?>, CapabilityMerger<?>> mergerTable = new HashMap<>();
        private final Set<Capability<?>> warned = new HashSet<>();

        @SuppressWarnings("NullableProblems")
        private Registry() {
            registerMerger(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, InventoryUtils::join);
            registerMerger(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, FluidHandlerConcatenate::new);
            registerMerger(CapabilityEnergy.ENERGY, EnergyUtils::join);
        }

        public <T> void registerMerger(Capability<T> capability, CapabilityMerger<T> merger) {
            if (mergerTable.containsKey(capability)) {
                CbtMod.LOGGER.warn("Ignoring duplicate capability merger {} for capability {} with existing merger {}",
                        merger.getClass().getCanonicalName(),
                        capability.getName(),
                        mergerTable.get(capability).getClass().getCanonicalName());
            } else {
                mergerTable.put(capability, merger);
            }
        }

        @SuppressWarnings("unchecked")
        @Nullable
        public <T> CapabilityMerger<T> getMerger(Capability<T> capability) {
            return (CapabilityMerger<T>)mergerTable.get(capability);
        }

        public <T> T merge(Capability<T> capability, List<T> capInstances) {
            CapabilityMerger<T> merger = getMerger(capability);
            if (merger == null) {
                CbtMod.LOGGER.warn("Capability {} has no registered merger; this could cause unexpected behaviour!",
                        capability.getName());
                return capInstances.get(0);
            }
            return merger.merge(capInstances);
        }

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/util/capability/CapabilityMerger.kt
package st.evening.mc.cbtweaker.util.capability

import it.unimi.dsi.fastutil.objects.ObjectOpenCustomHashSet
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.energy.CapabilityEnergy
import net.minecraftforge.fluids.capability.CapabilityFluidHandler
import net.minecraftforge.fluids.capability.templates.FluidHandlerConcatenate
import net.minecraftforge.items.CapabilityItemHandler
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.prelude.api.component.energy.ConcatEnergyStorage
import st.evening.mc.prelude.api.component.item.ConcatItemHandler
import st.evening.mc.prelude.api.util.collection.IdentityHashStrategy
import java.util.IdentityHashMap

fun interface CapabilityMerger<T : Any> {
    fun merge(instances: List<T>): T

    companion object {
        private val mergerTable: MutableMap<Capability<*>, CapabilityMerger<*>> = IdentityHashMap()
        private val warned: MutableSet<Capability<*>> = ObjectOpenCustomHashSet(IdentityHashStrategy())

        init {
            registerMerger(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, ConcatItemHandler::wrapMaybeModifiable)
            registerMerger(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, ::FluidHandlerConcatenate)
            registerMerger(CapabilityEnergy.ENERGY, ::ConcatEnergyStorage)
        }

        fun <T : Any> registerMerger(capability: Capability<T>, merger: CapabilityMerger<T>) {
            if (capability in mergerTable) {
                CbTweaker.logger.warn(
                    "Ignoring duplicate capability merger {} for capability {} with existing merger {}",
                    merger.javaClass.canonicalName,
                    capability.name,
                    mergerTable[capability]!!.javaClass.canonicalName
                )
            } else {
                mergerTable[capability] = merger
            }
        }

        @Suppress("UNCHECKED_CAST")
        fun <T : Any> getMerger(capability: Capability<T>): CapabilityMerger<T>? =
            mergerTable[capability] as? CapabilityMerger<T>

        fun <T : Any> merge(capability: Capability<T>, capInstances: List<T>): T {
            val merger = getMerger(capability)
            if (merger == null) {
                if (capability !in warned) {
                    warned += capability
                    CbTweaker.logger.warn(
                        "Capability {} has no registered merger; this could cause unexpected behaviour!",
                        capability.name
                    )
                }
                return capInstances[0]
            }
            return merger.merge(capInstances)
        }
    }
}
